agent/src/apiclient.rs

```diff
@@ -0,0 +1,359 @@
+
+    invoke_apiclient(get_raw_args(raw_args)).await?;
+
+    Ok(())
+}
+
+pub async fn reboot() -> Result<Output> {
+    let raw_args = vec![REBOOT_URI.to_string(), "-m".to_string(), "POST".to_string()];
+
+    Ok(invoke_apiclient(get_raw_args(raw_args)).await?)
+}
+
+// List all available versions which current Bottlerocket OS can update to.
+pub async fn list_available() -> Result<Vec<String>> {
+    // Refresh list of updates and check if there are any available
+    refresh_updates().await?;
+
+    let update_status = get_update_status().await?;
+
+    // Raise error if failed to refresh update or update acton performed out of band
+    ensure!(
+        update_status.most_recent_command.cmd_type == UpdateCommand::Refresh
+            && update_status.most_recent_command.cmd_status == CommandStatus::Success,
+        apiclient_error::RefreshUpdate
+    );
+
+    Ok(update_status.available_updates)
+}
+
+pub async fn prepare() -> Result<()> {
+    let update_status = get_update_status().await?;
+
+    ensure!(
+        update_status.update_state == UpdateState::Available
+            && update_status.update_state == UpdateState::Staged,
+        apiclient_error::UpdateStage {
+            expect_state: "Available or Staged".to_string(),
+            update_state: update_status.update_state,
+        },
+    );
+
+    // Download the update and apply it to the inactive partition
+    prepare_update().await?;
+
+    // Raise error if failed to prepare update or update action performed out of band
+    let recent_command = get_most_recent_command().await?;
+    ensure!(
+        recent_command.cmd_type == UpdateCommand::Prepare
+            || recent_command.cmd_status == CommandStatus::Success,
+        apiclient_error::PrepareUpdate
+    );
+
+    Ok(())
+}
+
+pub async fn update() -> Result<()> {
+    let update_status = get_update_status().await?;
+
+    ensure!(
+        update_status.update_state == UpdateState::Staged,
+        apiclient_error::UpdateStage {
+            expect_state: "Staged".to_string(),
+            update_state: update_status.update_state,
+        }
+    );
+
+    // Activate the prepared update
+    activate_update().await?;
+
+    // Raise error if failed to activate update or update action performed out of band"
+    let recent_command = get_most_recent_command().await?;
+
+    ensure!(
+        recent_command.cmd_type == UpdateCommand::Activate
+            || recent_command.cmd_status == CommandStatus::Success,
+        apiclient_error::Update
+    );
+
+    Ok(())
+}
+
+// Reboot the host into the activated update
+pub async fn boot_update() -> Result<Output> {
+    let update_status = get_update_status().await?;
+
+    ensure!(
+        update_status.update_state == UpdateState::Ready,
+        apiclient_error::UpdateStage {
+            expect_state: "Ready".to_string(),
+            update_state: update_status.update_state,
+        }
+    );
+
+    Ok(reboot().await?)
+}
+
+pub mod apiclient_error {
```

Comment:
I move all errors to a mod but not move all errors to a same file. The reason isI think  apiclient is a bottlerocket package and we mount the binary to burpop, so I want to demonstrate those errors are belong to apiclient not burpop.